src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs

```diff
@@ -0,0 +1,245 @@
+            ParseResponseHeaders(requestHandle, response, useDeflateDecompression || useGzipDecompression);
+
+            // Store response header cookies into custom CookieContainer.
+            if (cookieUsePolicy == CookieUsePolicy.UseSpecifiedCookieContainer)
+            {
+                Debug.Assert(cookieContainer != null);
+
+                if (response.Headers.Contains(HeaderNameSetCookie))
+                {
+                    IEnumerable<string> cookieHeaders = response.Headers.GetValues(HeaderNameSetCookie);
+                    foreach (var cookieHeader in cookieHeaders)
+                    {
+                        try
+                        {
+                            cookieContainer.SetCookies(request.RequestUri, cookieHeader);
+                        }
+                        catch (CookieException)
+                        {
+                            // We ignore malformed cookies in the response.
+                        }
+                    }
+                }
+            }
+
+            return response;
+        }
+
+        public static uint GetResponseHeaderNumberInfo(SafeWinHttpHandle requestHandle, uint infoLevel)
+        {
+            uint result = 0;
+            uint resultSize = sizeof(uint);
+
+            if (!Interop.WinHttp.WinHttpQueryHeaders(
+                requestHandle,
+                infoLevel | Interop.WinHttp.WINHTTP_QUERY_FLAG_NUMBER,
+                Interop.WinHttp.WINHTTP_HEADER_NAME_BY_INDEX,
+                ref result,
+                ref resultSize,
+                IntPtr.Zero))
+            {
+                WinHttpException.ThrowExceptionUsingLastError();
+            }
+
+            return result;
+        }
+
+        private static string GetResponseHeaderStringInfo(SafeWinHttpHandle requestHandle, uint infoLevel)
+        {
+            uint bytesNeeded = 0;
+            bool results = false;
+
+            // Call WinHttpQueryHeaders once to obtain the size of the buffer needed.  The size is returned in
+            // bytes but the API actually returns Unicode characters.
+            if (!Interop.WinHttp.WinHttpQueryHeaders(
+                requestHandle,
+                infoLevel,
+                Interop.WinHttp.WINHTTP_HEADER_NAME_BY_INDEX,
+                null,
+                ref bytesNeeded,
+                IntPtr.Zero))
+            {
+                int lastError = Marshal.GetLastWin32Error();
+                if (lastError == Interop.WinHttp.ERROR_WINHTTP_HEADER_NOT_FOUND)
+                {
+                    return null;
+                }
+
+                if (lastError != Interop.WinHttp.ERROR_INSUFFICIENT_BUFFER)
+                {
+                    throw WinHttpException.CreateExceptionUsingError(lastError);
+                }
+            }
+
+            // Allocate space for the buffer.
+            int charsNeeded = (int)bytesNeeded / 2;
+            var buffer = new StringBuilder(charsNeeded, charsNeeded);
```

Comment:
FYI, I did something similar in #3199 (among other changes), to avoid the `StringBuilder` overhead, but changed the p/invoke signature to take an `IntPtr` instead of `char[]` (to be consistent with how `IntPtr` was used instead of `byte[]` elsewhere in the WinHTTP p/invokes). It'd probably make more sense to just use `char[]` instead of `IntPtr` to avoid having to manually pin and pass the pointer to the `char[]` as an `IntPtr`.